pkg/operator/controllers/machineset/machineset_controller_test.go

```diff
@@ -0,0 +1,415 @@
+			r := &Reconciler{
+				log:    logrus.NewEntry(logrus.StandardLogger()),
+				arocli: arofake.NewSimpleClientset(&baseCluster),
+				maocli: tt.maocli,
+			}
+
+			request := ctrl.Request{}
+			request.Name = tt.objectName
+			request.Namespace = machineSetsNamespace
+			ctx := context.Background()
+
+			_, err := r.Reconcile(ctx, request)
+			if err != nil && err.Error() != tt.wantErr ||
+				err == nil && tt.wantErr != "" {
+				t.Error(err)
+			}
+
+			modifiedMachineset, err := tt.maocli.MachineV1beta1().MachineSets(request.Namespace).Get(ctx, request.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Error(err)
+			}
+
+			if *modifiedMachineset.Spec.Replicas != tt.wantReplicas {
+				t.Error(cmp.Diff(*modifiedMachineset.Spec.Replicas, tt.wantReplicas))
+			}
+		})
+	}
+}
+
+func TestFeatureFlag(t *testing.T) {
+	newFakeMao := func(replicas int32) *maofake.Clientset {
+		workerMachineSet := &machinev1beta1.MachineSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "aro-fake-machineset-0",
+				Namespace: machineSetsNamespace,
+				Labels: map[string]string{
+					"machine.openshift.io/cluster-api-machine-role": "worker",
+				},
+			},
+			Spec: machinev1beta1.MachineSetSpec{
+				Replicas: to.Int32Ptr(replicas),
+			},
+		}
+		return maofake.NewSimpleClientset(workerMachineSet)
+	}
+
+	baseCluster := arov1alpha1.Cluster{
+		ObjectMeta: metav1.ObjectMeta{Name: arov1alpha1.SingletonClusterName},
+		Spec: arov1alpha1.ClusterSpec{
+			InfraID:  "aro-fake",
+			Features: arov1alpha1.FeaturesSpec{ReconcileMachineSet: false},
+		},
+	}
+
+	tests := []struct {
+		name         string
+		maocli       *maofake.Clientset
+		wantReplicas int32
+		objectName   string
+		wantErr      string
+	}{
+		{
+			name:         "feature flag is false, replicas are incorrect",
+			maocli:       newFakeMao(0),
+			wantReplicas: 0,
+			wantErr:      "",
+			objectName:   "aro-fake-machineset-0",
+		},
+		{
+			name:         "feature flag is false, replicas are correct",
+			maocli:       newFakeMao(3),
+			wantReplicas: 3,
+			wantErr:      "",
+			objectName:   "aro-fake-machineset-0",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &Reconciler{
+				log:    logrus.NewEntry(logrus.StandardLogger()),
+				arocli: arofake.NewSimpleClientset(&baseCluster),
+				maocli: tt.maocli,
+			}
+
+			request := ctrl.Request{}
+			request.Name = tt.objectName
+			request.Namespace = machineSetsNamespace
+			ctx := context.Background()
+
+			_, err := r.Reconcile(ctx, request)
+			if err != nil && err.Error() != tt.wantErr ||
+				err == nil && tt.wantErr != "" {
+				t.Error(err)
+			}
+
+			modifiedMachineset, err := tt.maocli.MachineV1beta1().MachineSets(request.Namespace).Get(ctx, request.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Error(err)
+			}
+
+			if *modifiedMachineset.Spec.Replicas != tt.wantReplicas {
+				t.Error(cmp.Diff(*modifiedMachineset.Spec.Replicas, tt.wantReplicas))
+			}
+		})
```

Comment:
We discussed on a call that we should have single`TestReconciler` test to avoid copy-pasting this boilerplate.